### PR TITLE
[CDAP-15787] Add failure collector apis

### PR DESCRIPTION
**Pull request by CuriousVini**

### CODE REVIEW

File: adls-plugins/src/main/java/io/cdap/plugin/sink/ADLSBatchSink.java [hunk 1]
```diff
@@ -187,37 +195,43 @@ public AzureBatchSinkConfig(String referenceName) {
...
-      if (!containsMacro("path") && !path.startsWith("adl://")) {
-        throw new IllegalArgumentException("Path must start with adl:// for ADLS input files.");
+    protected void validate(FailureCollector collector) {
+      if (!containsMacro(PATH) && !path.startsWith("adl://")) {
+        collector.addFailure("Path must start with adl:// for ADLS input files.", null)
+          .withConfigProperty(PATH);
       }
       if ((AVRO.equals(outputFormat) || ORC.equals(outputFormat)) && !containsMacro("schema") && schema == null) {
-        throw new IllegalArgumentException("Output schema must be specified for avro or orc format output files.");
+        collector.addFailure("Output schema must be specified for avro or orc format output files.", null)
```
Comment (mikkin) [inline adls-plugins/src/main/java/io/cdap/plugin/sink/ADLSBatchSink.java]: nit: Capitalize acronyms

File: filesource-common/src/main/java/io/cdap/plugin/common/FileSourceConfig.java [hunk 2]
```diff
@@ -131,11 +136,16 @@ public FileSourceConfig(String referenceName, @Nullable String fileRegex, @Nulla
...
+  protected void validate(FailureCollector collector) {
+    try {
+      getFileSystemProperties();
+    } catch (IllegalArgumentException e) {
+      collector.addFailure(e.getMessage(), null).withConfigProperty(FILE_SYSTEM_PROPERTIES);
+    }
     if (!CombinePathTrackingInputFormat.class.getName().equals(inputFormatClass) && pathField != null) {
-      throw new IllegalArgumentException("pathField can only be used if inputFormatClass is " +
-                                           CombinePathTrackingInputFormat.class.getName());
+      collector.addFailure("Path field can only be used if inputFormatClass is " +
```
Comment (mikkin) [inline filesource-common/src/main/java/io/cdap/plugin/common/FileSourceConfig.java:146]: Curious if the `CombinePathTrackingInputFormat.class.getName()` adds a period at the end. Let me know if you have manually verified it.

**Review by mikkin — CHANGES_REQUESTED** [review]

File: adls-plugins/src/main/java/io/cdap/plugin/sink/ADLSBatchSink.java [hunk 3]
```diff
@@ -187,37 +195,43 @@ public AzureBatchSinkConfig(String referenceName) {
...
-            Schema inputFieldSchema = inputField.getSchema();
-            Schema.Type inputType = inputFieldSchema.isNullable() ? inputFieldSchema.getNonNullable().getType() :
-              inputFieldSchema.getType();
-            switch (inputType) {
-              case ARRAY:
-              case MAP:
-              case RECORD:
-                throw new IllegalArgumentException("Cannot include field with name " + fieldName + " of type " +
-                                                     inputType + " in output schema of text output format.");
+            collector.addFailure("Input schema does not contain the " + fieldName + " field.", null)
```
Comment (albertshau) [inline adls-plugins/src/main/java/io/cdap/plugin/sink/ADLSBatchSink.java]: enclose field name in single quotes like we do for the other messages.

File: adls-plugins/src/main/java/io/cdap/plugin/sink/ADLSBatchSink.java (hunk 1 above)
Comment (CuriousVini) [reply adls-plugins/src/main/java/io/cdap/plugin/sink/ADLSBatchSink.java]: this is how we display it on UI

**Review by shifuxu0301 — APPROVED** [review]
LGTM.

**Pull request merged**